### PR TITLE
fix(explorer): keep small Full Graph relationships readable

### LINKED ISSUE #1275: [BUG] Full Graph scatters small coordinate-free graphs and suppresses their edges
## Bug Description

Explorer's Full Graph view treats every coordinate-free graph as if it were a large graph.

For a small graph with several disconnected components, Explorer applies semantic-cluster seed spacing, starts ForceAtlas2 stabilization, and then applies full-overview edge LOD. The node and edge counters are correct, but the first useful view is not: components are spread across the canvas and ordinary relationships are hidden or too faint to read.

This is most visible with small extracted knowledge graphs, where seeing every relationship is more important than overview decluttering.

## Steps to Reproduce

Create a small graph with several connected components and no explicit coordinates:

```python
from semantica.context import ContextGraph

graph = ContextGraph()

nodes = [
    ("apple", "organization", "Apple Inc."),
    ("jobs", "person", "Steve Jobs"),
    ("wozniak", "person", "Steve Wozniak"),
    ("wayne", "person", "Ronald Wayne"),
    ("cupertino", "place", "Cupertino"),
    ("california", "place", "California"),
    ("person", "class", "Person"),
    ("place", "class", "Place"),
    ("name", "property", "name"),
    ("metadata", "property", "metadata"),
    ("string", "datatype", "xsd:string"),
    ("ontology", "ontology", "ExampleOntology"),
]

for node_id, node_type, content in nodes:
    graph.add_node(node_id, node_type, content=content)

edges = [
    ("apple", "jobs", "founded_by"),
    ("jobs", "apple", "works_for"),
    ("wayne", "cupertino", "located_in"),
    ("name", "person", "rdfs:domain"),
    ("name", "place", "rdfs:domain"),
    ("name", "string", "rdfs:range"),
    ("metadata", "person", "rdfs:domain"),
    ("metadata", "place", "rdfs:domain"),
    ("metadata", "string", "rdfs:range"),
]

for source, target, edge_type in edges:
    graph.add_edge(source, target, edge_type)

graph.save_to_file("small-graph-repro.json")
```

Build and start Explorer:

```bash
cd explorer
npm ci
npm run build
cd ..

python -m semantica.explorer \
  --graph small-graph-repro.json \
  --port 8001 \
  --no-browser
```

Open `http://127.0.0.1:8001/`, select Knowledge Explorer, and keep the default Full Graph view.

## Expected Behavior

For a small graph without usable coordinates:

- connected nodes are placed near one another;
- disconnected components are compactly packed into the initial viewport;
- positions are deterministic for the same graph regardless of input order;
- node labels and all nine relationship lines remain readable at overview zoom;
- the graph does not run an unnecessary force-stabilization phase.

Graphs with usable coordinates and graphs above the small-graph threshold should keep the existing behavior.

## Actual Behavior

- The header correctly reports `12 nodes / 9 edges`.
- Several nodes appear far away from the components they belong to.
- Ordinary edges are classified as inactive/hidden by full-view overview LOD.
- Relationship labels are missing or too faint to establish the graph's structure.
- ForceAtlas2 runs even though a compact deterministic layout is sufficient.

## Root Cause

Two large-graph policies are applied without considering graph size:

1. In `useLoadGraph.ts`, any graph with insufficient coordinate coverage receives semantic-cluster seed positions and `layoutReady: false`, which triggers runtime force layout.
2. In `graphSceneState.ts`, Full Graph overview priority thresholds, muted-state hiding, and background sampling suppress ordinary edges.

Those policies are useful for dense graphs but counterproductive for a graph with only a few dozen nodes.

## Why This Matters

The Explorer API and store contain the correct relationships, while the visual result makes them appear absent. This can be mistaken for extraction or ingestion data loss and makes small cookbook or onboarding graphs difficult to debug.

## Environment

- Semantica: `0.6.7`, current `main` at `dfda4c56`
- Python: `3.12.4`
- Node.js: `24.14.0`
- npm: `11.9.0`
- OS used for reproduction: macOS 26.6.2
- Installation: source checkout with a production Explorer build

## Acceptance Criteria

- [ ] Small coordinate-free graphs use a compact deterministic layout.
- [ ] Connected components remain visually grouped.
- [ ] Small-graph node labels are visible in the initial overview.
- [ ] Small-graph relationships bypass large-graph priority hiding and sampling.
- [ ] Force stabilization is skipped when the deterministic layout is complete.
- [ ] Input ordering does not change the resulting positions.
- [ ] Graphs with usable coordinates are unchanged.
- [ ] Graphs above the small-graph threshold retain the existing large-graph behavior.
- [ ] Unit tests cover layout selection, deterministic component packing, and edge visibility.

## Possible Solution

Select a dedicated small-graph layout only when node count is below a bounded threshold and coordinate coverage is insufficient. Pack connected components deterministically, mark the layout ready immediately, and carry a small-graph display hint so Full Graph overview preserves those edges without changing large-graph LOD.

### PR DESCRIPTION
**Pull request by taoche**
## Description

Explorer's Full Graph view currently applies the large-graph seed layout, ForceAtlas2 stabilization, and overview edge LOD to every coordinate-free graph. On a small graph with several disconnected components, that makes nodes appear as unlabeled dots and suppresses ordinary relationships at the initial overview zoom.

This change adds a deterministic, component-aware layout path for coordinate-free graphs with at most 48 nodes. Small graphs skip unnecessary force stabilization, keep node labels visible, and retain readable relationship edges. Existing behavior is unchanged for larger graphs and for graphs that already provide usable coordinates.

## Before / After

Both screenshots use the same 1280×720 viewport and the same fixture: 12 nodes, 9 edges, 3 connected components, and no node coordinates.

| Before (`main`) | After (this PR) |
| --- | --- |
| Before: nodes are tiny and most relationships are unreadable | After: compact components with readable node and relationship labels |

## Type of Change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature which would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Performance improvement
- [ ] Code refactoring

## Related Issues

Closes #1275

## Root Cause

- Coordinate-free small graphs used the same cluster seed spacing and ForceAtlas2 stabilization as large graphs.
- Full Graph overview LOD sampled or hid ordinary low-priority edges even when the entire graph contained only a few relationships.
- Default node sizes and label policy were tuned for large-graph density rather than small-graph readability.

## Changes Made

- Select a deterministic small-graph layout only for graphs with 1–48 nodes and less than 92% coordinate coverage.
- Discover connected components deterministically and pack them into a compact centered grid.
- Mark the layout ready immediately so small graphs do not run ForceAtlas2 stabilization.
- Use readable minimum node sizes and always-visible labels for the small-graph path.
- Preserve ordinary small-graph relationship edges through overview LOD while retaining the existing large-graph policy.
- Add focused regression tests for selection, determinism, component proximity, and low-priority edge visibility.

## Reproduction

1. Load a graph containing roughly 10–20 nodes, multiple disconnected components, and no `x`/`y` node coordinates.
2. Open Knowledge Explorer → Full Graph.
3. Compare the initial overview without selecting or hovering a node.

Before this change, the graph spends time stabilizing and displays mostly tiny unlabeled points with relationships hidden or too faint to read. After this change, the initial view is compact and deterministic, and node/relationship labels are visible.

## Testing

- [x] Tested locally
- [x] Added tests for new functionality
- [x] Package builds successfully (`python -m build`)

### Test Commands

```bash
cd explorer
npm run test:graph-workspace  # 81 passed
npm run build                 # production build succeeded

cd ..
python -m build               # sdist and wheel built successfully
```

## Documentation

- [ ] Updated relevant documentation
- [ ] Added code examples if applicable
- [ ] Updated API reference if adding new APIs
- [ ] Updated cookbook if adding new examples
- [x] No documentation changes needed

The screenshots above are GitHub-hosted PR attachments and are not part of the repository diff.

## Breaking Changes

**Breaking Changes**: No

## Checklist

- [x] My code follows the project's style guidelines
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] Package builds successfully

## Additional Notes

The small-graph threshold and coordinate-coverage threshold are isolated in `smallGraphLayout.ts`. Graphs above the threshold continue to use the existing large-graph layout and edge LOD behavior.

### CODE REVIEW

File: explorer/src/workspaces/GraphWorkspace/useLoadGraph.ts [hunk 1]
```diff
@@ -554,9 +558,15 @@ export function useLoadGraph(options: UseLoadGraphOptions = {}) {
...
       const coordinateCoverage = fetchedNodes.length > 0 ? providedCoordinateCount / fetchedNodes.length : 0;
       const useProvidedCoordinates = coordinateCoverage >= 0.92;
+      const useSmallGraphLayout = shouldUseSmallGraphLayout(fetchedNodes.length, coordinateCoverage);
       const seededPositions = useProvidedCoordinates
         ? null
-        : buildClusterSeedPositions(
+        : useSmallGraphLayout
+          ? buildSmallGraphSeedPositions(
+              fetchedNodes.map((node) => node.id),
+              fetchedEdges,
```
[the remark this answers is not in the document]
Comment (taoche) [reply explorer/src/workspaces/GraphWorkspace/useLoadGraph.ts:571]: Fixed in 50ddbc06. Eligible small graphs now consistently use deterministic seed positions for both axes; partial/far-away provided coordinates are ignored on the compact-layout path. Added a regression covering full and one-axis partial outlier coordinates.

File: explorer/tests/smallGraphLayout.test.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+import {
+  SMALL_GRAPH_MAX_NODES,
+  buildSmallGraphSeedPositions,
+  shouldUseSmallGraphLayout,
+} from "../src/workspaces/GraphWorkspace/smallGraphLayout.ts";
+
+test("small graph layout is selected only when coordinates are not already usable", () => {
+  assert.equal(shouldUseSmallGraphLayout(12, 0), true);
+  assert.equal(shouldUseSmallGraphLayout(SMALL_GRAPH_MAX_NODES + 1, 0), false);
+  assert.equal(shouldUseSmallGraphLayout(12, 0.95), false);
```
[the remark this answers is not in the document]
Comment (taoche) [reply explorer/tests/smallGraphLayout.test.ts:15]: Fixed in 50ddbc06. The load path now uses one resolveGraphLayoutDecision result for coordinate selection and the returned layoutReady summary. Regression tests assert eligible small graphs return layoutReady=true (so GraphWorkspace does not start runtime stabilization), while above-threshold graphs remain layoutReady=false and coordinate-aware graphs remain ready.

File: explorer/src/workspaces/GraphWorkspace/graphSceneState.ts [hunk 3]
```diff
@@ -1783,18 +1783,21 @@ export function resolveEdgeElementStyle(
   const isCommunityBundle = attrs.bundleKind === "community";
   const baseSize = Number(attrs.baseSize || attrs.size || 0.9);
   const visualPriority = Number(attrs.visualPriority ?? 0);
+  const isSmallGraphEdge = viewMode === "full" && attrs.isSmallGraph === true;
```
[the remark this answers is not in the document]
Comment (taoche) [reply explorer/src/workspaces/GraphWorkspace/graphSceneState.ts:1786]: Fixed in 50ddbc06. Small-graph display mode is now held as a graph-level ref, realtime ADD_EDGE attributes receive that state, and all edge markers are synchronized after ADD_NODE/ADD_EDGE (including disabling the marker when the graph exceeds 48 nodes). Added focused realtime attribute regressions for both enabled and disabled states.

File: explorer/src/workspaces/GraphWorkspace/smallGraphLayout.ts [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    });
+    components.push(component);
+  });
+
+  components.sort((left, right) => right.length - left.length || left[0].localeCompare(right[0]));
+
+  const columns = Math.max(1, Math.ceil(Math.sqrt(components.length)));
+  const rows = Math.max(1, Math.ceil(components.length / columns));
+  const cellWidth = 190;
+  const cellHeight = 150;
```
[the remark this answers is not in the document]
Comment (taoche) [reply explorer/src/workspaces/GraphWorkspace/smallGraphLayout.ts]: Fixed in 50ddbc06. Grid spacing is now 204 units: two maximum 78-unit radii plus a 48-unit component gap. Added a 48-node/four-component regression that checks the minimum pairwise distance between every disconnected component.

**Review by Sameer6305 — APPROVED** [review]
Thanks for the solid contribution! 🙌

This PR addresses #1275, fixing the Full Graph experience for small coordinate-free graphs by introducing a deterministic, component-aware layout for graphs up to 48 nodes. It keeps connected components compact, makes labels and relationships readable, skips unnecessary ForceAtlas2 stabilization, and preserves the existing behavior for coordinate-backed and larger graphs.

What looks solid:

Deterministic component discovery and packing.
Small-graph threshold and coordinate-coverage handling are isolated.
Small-graph edges correctly bypass the large-graph overview LOD.
Existing large-graph and coordinate-backed behavior remains unchanged.
Regression coverage was added for layout selection, determinism, component proximity, realtime edges, and edge visibility.

We also investigated the CI concern and verified the Explorer test suite/build are passing (81/81 tests, production build successful). The earlier CI issue was not caused by the implementation itself.

During final verification, we also pushed a small performance improvement: removed the redundant synchronizeRealtimeSmallGraphEdges() call after every realtime edge addition. The new edge already receives the correct isSmallGraph attribute, while the required synchronization on the small → large node transition remains intact.

We verified the final change with tests, production build, and git diff --check.

Everything looks correct and merge-ready. Approved. ✅
